File: android/app/src/main/kotlin/jp/golia/mailrs/accounts/ImapSession.kt
package jp.golia.mailrs.accounts

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.Writer
import java.net.Socket
import javax.net.ssl.SSLSocket
import javax.net.ssl.SSLSocketFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * A conversation with an IMAP server.
 *
 * The socket half. Everything it reads is handed to [Imap] to
 * interpret, so this file is about bytes and timeouts and that file is
 * about grammar — and only one of them needs a server to test.
 *
 * **TLS from the first byte.** No plaintext, and no STARTTLS: a
 * credential is not sent over a connection that was ever in the clear,
 * and every provider worth connecting to offers 993.
 */
class ImapSession(private val host: String, private val port: Int) : AutoCloseable {
    sealed class Failure(message: String) : Exception(message) {
        /** Could not reach the server at all. */
        class Unreachable(val why: String) : Failure(why)

        /**
         * The credential was refused. **Not a network problem** — one
         * is a button to press, the other is waiting.
         */
        class Refused(val detail: String) : Failure(detail)

        /** The server said no to something else. */
        class Server(val detail: String) : Failure(detail)

        /** The connection broke mid-conversation. */
        class Closed : Failure("the server closed the connection")
    }

    /**
     * Where the bytes go.
     *
     * A seam, so the conversation above it can be tested without a
     * server. Every rule this session enforces — a tag prefix that must
     * not match `a10` when it asked as `a1`, a literal read by the byte
     * count the server announced, a folder name at the end of a LIST
     * line — is a rule about what arrives, and none of them can be
     * checked by connecting to a real server and hoping it sends the
     * awkward case.
     */
    interface Transport {
        fun readLine(): String
        fun readBytes(count: Int): String
        fun write(text: String)
        fun close()
    }

    /** Swapped in by tests; null means the real socket. */
    internal var transport: Transport? = null

    private var socket: SSLSocket? = null
    private var reader: BufferedReader? = null
    private var writer: Writer? = null
    private var tag = 0

    /** Open the connection and read the greeting. */
    suspend fun connect(timeoutMs: Int = 20_000) = withContext(Dispatchers.IO) {
        // A scripted server has no socket to open. This branch was
        // missing at first and nothing noticed: the unit tests call
        // `list` and `login` directly and never `connect`, so the seam
        // had a hole exactly where no test crossed it — and the first
        // end-to-end run went off and dialled localhost:993.
        if (transport != null) {
            val greeting = readLine()
            if (greeting.uppercase().startsWith("* BYE")) {
                throw Failure.Server(greeting.removePrefix("* "))
            }
            return@withContext
        }
        try {
            val plain = Socket()
            plain.connect(java.net.InetSocketAddress(host, port), timeoutMs)
            plain.soTimeout = 60_000
            val s = (SSLSocketFactory.getDefault() as SSLSocketFactory)
                .createSocket(plain, host, port, true) as SSLSocket
            // Names are checked against the certificate. Without this
            // any certificate a machine on the path can produce is
            // accepted, which is the whole of the protection gone.
            s.sslParameters = s.sslParameters.apply { endpointIdentificationAlgorithm = "HTTPS" }
            s.startHandshake()
            socket = s
            // **ISO-8859-1, deliberately.** It maps every byte to the
            // code point of the same value, so what is read back is
            // exactly what arrived and nothing has been decided yet. A
            // UTF-8 reader here settles a message's charset before the
            // header declaring it has been read, and every Shift_JIS
            // and windows-1252 body arrives as replacement characters.
            // `Wire.utf8` turns it back into text where text is what is
            // wanted.
            reader = BufferedReader(InputStreamReader(s.inputStream, Charsets.ISO_8859_1))
            writer = s.outputStream.writer(Charsets.UTF_8)
        } catch (e: Exception) {
            throw Failure.Unreachable(e.message ?: e.toString())
        }
        val greeting = readLine()
        if (greeting.uppercase().startsWith("* BYE")) {
            throw Failure.Server(greeting.removePrefix("* "))
        }
    }

    override fun close() {
        transport?.let {
            runCatching { it.close() }
            return
        }
        runCatching { socket?.close() }
        socket = null
    }

    /**
     * Sign in.
     *
     * The password is quoted: generated app passwords contain `"` and
     * `\` often enough that an unquoted LOGIN turns one into a syntax
     * error — and the person is told their password is wrong when it
     * is right.
     */
    suspend fun login(user: String, password: String) = withContext(Dispatchers.IO) {
        val (_, done) = command("LOGIN ${Imap.quoted(user)} ${Imap.quoted(password)}")
        when (done) {
            is Imap.Completion.Ok -> Unit
            is Imap.Completion.No ->
                throw if (Imap.isAuthenticationFailure(done.detail)) {
                    Failure.Refused(done.detail)
                } else {
                    Failure.Server(done.detail)
                }
            is Imap.Completion.Bad -> throw Failure.Server(done.detail)
        }
    }

    /**
     * Sign in with an access token.
     *
     * A provider that rejects the token answers a continuation (`+`)
     * with a base64 error rather than a final code, and will not send
     * one until the client sends an empty line. Reading that `+` as
     * success authenticates every refused token.
     */
    suspend fun authenticateXOAuth2(user: String, token: String) = withContext(Dispatchers.IO) {
        val t = nextTag()
        write("$t AUTHENTICATE XOAUTH2 ${Smtp.authXOAuth2(user, token)}\r\n")
        while (true) {
            val line = readLine()
            if (line.startsWith("+")) {
                write("\r\n")
                continue
            }
            when (val done = Imap.completion(line, t)) {
                is Imap.Completion.Ok -> return@withContext
                is Imap.Completion.No -> throw Failure.Refused(done.detail)
                is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                null -> Unit
            }
        }
    }

    /** Every folder the server offers. */
    suspend fun list(): List<Imap.Untagged.ListFolder> = withContext(Dispatchers.IO) {
        val (untagged, done) = command("""LIST "" "*"""")
        refuseIfNotOk(done)
        untagged.filterIsInstance<Imap.Untagged.ListFolder>()
    }

    /** Open a folder, and say what state it is in. */
    suspend fun select(folder: String): Pair<Long, Int> = withContext(Dispatchers.IO) {
        val (untagged, done) = command("SELECT ${Imap.quoted(folder)}")
        refuseIfNotOk(done)
        var validity = 0L
        var exists = 0
        for (u in untagged) {
            when (u) {
                is Imap.Untagged.UidValidity -> validity = u.value
                is Imap.Untagged.Exists -> exists = u.count
                else -> Unit
            }
        }
        validity to exists
    }

    private fun refuseIfNotOk(done: Imap.Completion) {
        when (done) {
            is Imap.Completion.No -> throw Failure.Server(done.detail)
            is Imap.Completion.Bad -> throw Failure.Server(done.detail)
            is Imap.Completion.Ok -> Unit
        }
    }

    /** One message, as far as a list row needs it. */
    data class Fetched(
        val uid: Long,
        val seen: Boolean,
        val headers: MessageHeaders.Parsed,
        /**
         * Seconds since the epoch, or null when the `Date:` header
         * could not be read. **Null, not now** — a message shown as
         * having just arrived jumps to the top and stays there.
         */
        val date: Long?,
        /** How big the whole message is, when the server said. */
        val size: Long? = null,
    )

    /**
     * Read the headers of everything in [range].
     *
     * `BODY.PEEK[HEADER]`, not `BODY[HEADER]`: the second marks every
     * message read on the server just for having been listed, which is
     * the rudest thing a mail client can do to somebody's mailbox.
     *
     * The literal is read **by the byte count the server announced**,
     * never by scanning for a terminator: a message contains every
     * byte sequence a terminator could be made of.
     */
    /**
     * @param byUid whether [range] is uids or **positions**. `UID
     *   FETCH 1:500` and `FETCH 1:500` mean completely different
     *   things, and a first pass counts from the end of the folder,
     *   which only a position can say.
     */
    suspend fun fetchHeaders(range: String, byUid: Boolean = true): List<Fetched> =
        withContext(Dispatchers.IO) {
        val t = nextTag()
        val verb = when {
            byUid -> "UID FETCH"
            else -> "FETCH"
        }
        write("$t $verb $range (UID FLAGS RFC822.SIZE BODY.PEEK[HEADER])\r\n")
        val out = mutableListOf<Fetched>()
        while (true) {
            val line = readLine()
            Imap.completion(line, t)?.let { done ->
                when (done) {
                    is Imap.Completion.Ok -> return@withContext out
                    is Imap.Completion.No -> throw Failure.Server(done.detail)
                    is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                }
            }
            val announced = Imap.fetchLine(line) ?: continue
            val uid = announced.uid ?: continue
            // A flags-only reply: nothing to read, and nothing a row
            // can show that it does not already have.
            val count = announced.literalBytes ?: continue
            val headers = MessageHeaders.parse(Wire.utf8(readBytes(count)))
            out += Fetched(
                uid, announced.seen, headers, MailDate.epochSeconds(headers.date),
                announced.size,
            )
        }
        @Suppress("UNREACHABLE_CODE")
        out
    }

    /**
     * One message, whole.
     *
     * `BODY.PEEK[]` rather than `BODY[]` for the same reason the header
     * fetch peeks: opening a message is the reader's decision, and a
     * client that marks mail read for having looked at it takes that
     * decision away. Marking read is a separate, deliberate call.
     */
    /**
     * @param plan whether to take the whole message or only its
     *   beginning. A message with a 25 MB attachment is 25 MB to
     *   fetch, and fetching it to show two lines of text is noticed on
     *   a bill rather than on a screen.
     */
    suspend fun fetchRaw(
        uid: Long,
        plan: FetchWhole.Plan = FetchWhole.Plan.Whole,
    ): ByteArray = withContext(Dispatchers.IO) {
        val t = nextTag()
        write("$t UID FETCH $uid (${FetchWhole.bodyItem(plan)})\r\n")
        var out = ByteArray(0)
        while (true) {
            val line = readLine()
            Imap.completion(line, t)?.let { done ->
                when (done) {
                    is Imap.Completion.Ok -> return@withContext out
                    is Imap.Completion.No -> throw Failure.Server(done.detail)
                    is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                }
            }
            val announced = Imap.fetchLine(line) ?: continue
            val count = announced.literalBytes ?: continue
            // The announced byte count, never a scan for a terminator: a
            // message contains every byte sequence a terminator could be
            // made of.
            out = Wire.bytes(readBytes(count))
        }
        @Suppress("UNREACHABLE_CODE")
        out
    }

    /**
     * Ask the server for nothing, to find out whether it is still
     * there.
     *
     * A connection kept for reuse may have been dropped since — by an
     * idle timeout, by a NAT, by a server restart — and the socket
     * gives no sign of it until something is written. This is that
     * something, and it costs one round trip rather than a whole
     * reconnection.
     */
    suspend fun noop() = withContext(Dispatchers.IO) {
        val (_, done) = command("NOOP")
        refuseIfNotOk(done)
    }

    /** Mark a message read on the server, because somebody read it. */
    suspend fun markSeen(uid: Long) = store(uid, "+FLAGS", "\\Seen")

    /**
     * Mark it unread again.
     *
     * `-FLAGS`, not a `FLAGS` that names what should remain: the
     * second replaces the whole set, so it would quietly clear
     * `\Flagged`, `\Answered` and every keyword the person or another
     * client had put there.
     */
    suspend fun markUnseen(uid: Long) = store(uid, "-FLAGS", "\\Seen")

    private suspend fun store(uid: Long, op: String, flag: String) = withContext(Dispatchers.IO) {
        val t = nextTag()
        write("$t UID STORE $uid $op ($flag)\r\n")
        while (true) {
            val line = readLine()
            Imap.completion(line, t)?.let { done ->
                when (done) {
                    is Imap.Completion.Ok -> return@withContext
                    is Imap.Completion.No -> throw Failure.Server(done.detail)
                    is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                }
            }
        }
    }

    /**
     * The flags of messages already held, and which of them are gone.
     *
     * A sync that only fetches new uids never notices what happened
     * elsewhere: a message read on a laptop stays bold here forever,
     * and one deleted there stays in the list forever. Both are things
     * a person sees immediately and cannot explain.
     *
     * Cheap on purpose — flags only, for uids this device already has,
     * so the answer is a few bytes per message rather than a header
     * block.
     *
     * @return uid to seen, for every uid the server still has. **A uid
     *   that was asked about and is missing from the answer is one the
     *   server no longer has**, which is how deletion is noticed at
     *   all.
     */
    suspend fun flags(uids: List<Long>): Map<Long, Boolean> = withContext(Dispatchers.IO) {
        val out = mutableMapOf<Long, Boolean>()
        // Collapsed to ranges and split into commands a server will
        // accept: naming five thousand uids one by one is a line tens
        // of kilobytes long, and the mailbox that most needs its flags
        // refreshed would be the one where the refresh stops working.
        for (batch in UidRanges.batches(uids)) {
            val t = nextTag()
            write("$t UID FETCH $batch (UID FLAGS)\r\n")
            while (true) {
                val line = readLine()
                val done = Imap.completion(line, t)
                if (done != null) {
                    when (done) {
                        is Imap.Completion.Ok -> Unit
                        is Imap.Completion.No -> throw Failure.Server(done.detail)
                        is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                    }
                    break
                }
                val announced = Imap.fetchLine(line) ?: continue
                val uid = announced.uid ?: continue
                out[uid] = announced.seen
            }
        }
        out
    }

    /**
     * What the server said it can do.
     *
     * Read rather than assumed: the two commands below exist only on
     * some servers, and asking for one that is not there is an error
     * the person sees rather than a fallback they do not.
     */
    suspend fun capabilities(): Set<String> = withContext(Dispatchers.IO) {
        val (untagged, done) = command("CAPABILITY")
        refuseIfNotOk(done)
        untagged.filterIsInstance<Imap.Untagged.Capabilities>()
            .flatMap { it.names }
            .map { it.uppercase() }
            .toSet()
    }

    /** Put a message in another folder. See [MovePlan]. */
    suspend fun moveTo(uid: Long, folder: String, capabilities: Set<String>) =
        withContext(Dispatchers.IO) {
            for (step in MovePlan.steps(uid, folder, capabilities)) {
                when (step) {
                    is MovePlan.Step.MarkDeleted -> store(uid, "+FLAGS", "\\Deleted")
                    is MovePlan.Step.Command -> {
                        val t = nextTag()
                        write("$t ${step.text}\r\n")
                        awaitCompletion(t)
                    }
                }
            }
        }

    private suspend fun awaitCompletion(tag: String) = withContext(Dispatchers.IO) {
        while (true) {
            val line = readLine()
            Imap.completion(line, tag)?.let { done ->
                when (done) {
                    is Imap.Completion.Ok -> return@withContext
                    is Imap.Completion.No -> throw Failure.Server(done.detail)
                    is Imap.Completion.Bad -> throw Failure.Server(done.detail)
                }
            }
        }
    }

    /** Exactly [count] bytes, whatever they contain. */
    private fun readBytes(count: Int): String {
        transport?.let { return it.readBytes(count) }
        return readBytesFromSocket(count)
    }

    private fun readBytesFromSocket(count: Int): String {
        val buf = CharArray(count)
        var read = 0
        val r = reader ?: throw Failure.Closed()
        while (read < count) {
            val n = r.read(buf, read, count - read)
            if (n < 0) throw Failure.Closed()
            read += n
        }
        return String(buf, 0, read)
    }

    // MARK: the wire

    private fun nextTag(): String = "a${++tag}"

    /** Send a command and read until its tagged reply. */
    private fun command(text: String): Pair<List<Imap.Untagged>, Imap.Completion> {
        val t = nextTag()
        write("$t $text\r\n")
        val untagged = mutableListOf<Imap.Untagged>()
        while (true) {
            val line = readLine()
            Imap.completion(line, t)?.let { return untagged to it }
            Imap.untagged(line)?.let { untagged += it }
        }
    }

    private fun write(text: String) {
        transport?.let {
            it.write(text)
            return
        }
        val w = writer ?: throw Failure.Closed()
        w.write(text)
        w.flush()
    }

    private fun readLine(): String {
        transport?.let { return it.readLine() }
        return reader?.readLine() ?: throw Failure.Closed()
    }
}

File: ios/Mailrs/Accounts/IMAPSession.swift
import Foundation
import Network

/// A conversation with an IMAP server.
///
/// The socket half. Everything it reads is handed to `IMAP` to
/// interpret, so this file is about bytes and timeouts and that file
/// is about grammar — and only one of them needs a server to test.
///
/// **TLS from the first byte.** No plaintext, and no STARTTLS: a
/// credential is not sent over a connection that was ever in the
/// clear, and every provider worth connecting to offers 993.
actor IMAPSession {
    enum Failure: Error, Equatable {
        /// Could not reach the server at all.
        case unreachable(String)
        /// The credential was refused. **Not a network problem** — one
        /// is a button to press, the other is waiting.
        case refused(String)
        /// The server said no to something else.
        case server(String)
        /// The connection broke mid-conversation.
        case closed
        /// The server did not answer in time.
        case timedOut
    }

    private let transport: ByteTransport
    private var tag = 0
    private var buffer = Data()
    /// How long any one command may take before it is a timeout.
    ///
    /// Generous, because a large FETCH on a slow link is not a fault.
    private let commandTimeout: Duration = .seconds(60)

    init(host: String, port: UInt16) {
        transport = TLSTransport(host: host, port: port)
    }

    /// For tests: a scripted server, and the conversation under test.
    ///
    /// Every rule this session enforces — a tag that must not be
    /// matched by its own prefix, a literal read by the byte count the
    /// server announced, a folder name at the end of a LIST line — is
    /// a rule about what arrives, and a real server will not send the
    /// awkward case on demand.
    init(transport: ByteTransport) {
        self.transport = transport
    }

    /// Open the connection and read the greeting.
    func connect() async throws {
        do {
            try await transport.connect()
        } catch let e as TransportFailure {
            throw failure(e)
        }
        // The greeting. A server that refuses the connection says so
        // here — `* BYE`, often with a reason worth showing.
        let greeting = try await readLine()
        if greeting.uppercased().hasPrefix("* BYE") {
            throw Failure.server(String(greeting.dropFirst(2)))
        }
    }

    func close() async {
        await transport.close()
    }

    private func failure(_ e: TransportFailure) -> Failure {
        switch e {
        case let .unreachable(why): return .unreachable(why)
        case .closed: return .closed
        case .cannotUpgrade: return .unreachable("this connection could not be encrypted")
        }
    }

    /// Sign in.
    ///
    /// The password is quoted: generated app passwords contain `"` and
    /// `\` often enough that an unquoted LOGIN turns one into a syntax
    /// error — and the person is told their password is wrong when it
    /// is right.
    func login(user: String, password: String) async throws {
        let (_, completion) = try await command(
            "LOGIN \(IMAP.quoted(user)) \(IMAP.quoted(password))")
        switch completion {
        case .ok:
            return
        case let .no(detail):
            throw IMAP.isAuthenticationFailure(detail)
                ? Failure.refused(detail) : Failure.server(detail)
        case let .bad(detail):
            throw Failure.server(detail)
        }
    }

    /// Sign in with an access token.
    ///
    /// An access token is not a password: `\u{1}` separators, an
    /// `auth=Bearer ` prefix, two terminators — and a different
    /// failure protocol. A provider that rejects the token answers a
    /// continuation (`+`) with a base64 error rather than a final
    /// code, and will not send one until the client sends an empty
    /// line. Reading that `+` as success authenticates every refused
    /// token.
    func authenticateXOAuth2(user: String, token: String) async throws {
        let raw = "user=\(user)\u{1}auth=Bearer \(token)\u{1}\u{1}"
        let payload = Data(raw.utf8).base64EncodedString()
        let t = nextTag()
        try await send("\(t) AUTHENTICATE XOAUTH2 \(payload)\r\n")
        while true {
            let line = try await readLine()
            if line.hasPrefix("+") {
                // Refused. Answer the continuation so the server sends
                // the tagged failure rather than waiting for us.
                try await send("\r\n")
                continue
            }
            if let completion = IMAP.completion(of: line, tag: t) {
                switch completion {
                case .ok: return
                case let .no(d): throw Failure.refused(d)
                case let .bad(d): throw Failure.server(d)
                }
            }
        }
    }

    /// Every folder the server offers.
    func list() async throws -> [(name: String, attributes: [String])] {
        let (untagged, completion) = try await command(#"LIST "" "*""#)
        if case let .no(d) = completion { throw Failure.server(d) }
        if case let .bad(d) = completion { throw Failure.server(d) }
        return untagged.compactMap {
            if case let .list(name, attributes) = $0 { return (name, attributes) }
            return nil
        }
    }

    /// Open a folder, and say what state it is in.
    func select(_ folder: String) async throws -> (uidValidity: UInt32, exists: Int) {
        let (untagged, completion) = try await command("SELECT \(IMAP.quoted(folder))")
        if case let .no(d) = completion { throw Failure.server(d) }
        if case let .bad(d) = completion { throw Failure.server(d) }
        var validity: UInt32 = 0
        var exists = 0
        for u in untagged {
            if case let .uidValidity(v) = u { validity = v }
            if case let .exists(n) = u { exists = n }
        }
        return (validity, exists)
    }

    /// One message, as far as a list row needs it.
    struct Fetched: Equatable {
        let uid: UInt32
        let seen: Bool
        let headers: MessageHeaders.Parsed
        /// Seconds since the epoch, or nil when the `Date:` header
        /// could not be read. **Nil, not now** — a message shown as
        /// having just arrived jumps to the top and stays there.
        let date: Int64?
    }

    /// Read the headers of everything in `range`.
    ///
    /// `BODY.PEEK[HEADER]`, not `BODY[HEADER]`: the second marks every
    /// message read on the server just for having been listed, which
    /// is the rudest thing a mail client can do to somebody's mailbox.
    ///
    /// The literal is read **by the byte count the server announced**,
    /// never by scanning for a terminator: a message contains every
    /// byte sequence a terminator could be made of.
    func fetchHeaders(range: String) async throws -> [Fetched] {
        let t = nextTag()
        try await send("\(t) UID FETCH \(range) (UID FLAGS BODY.PEEK[HEADER])\r\n")
        var out: [Fetched] = []
        while true {
            let line = try await readLine()
            if let done = IMAP.completion(of: line, tag: t) {
                switch done {
                case .ok: return out
                case let .no(d): throw Failure.server(d)
                case let .bad(d): throw Failure.server(d)
                }
            }
            guard let announced = IMAP.fetchLine(line), let uid = announced.uid else { continue }
            guard let count = announced.literalBytes else {
                // A flags-only reply: nothing to read, and nothing a
                // row can show that it does not already have.
                continue
            }
            let raw = try await readBytes(count)
            let headers = MessageHeaders.parse(raw)
            out.append(
                Fetched(
                    uid: uid,
                    seen: announced.seen,
                    headers: headers,
                    date: MailDate.epochSeconds(headers.date)))
        }
    }

    /// Exactly `count` bytes, whatever they contain.
    private func readBytes(_ count: Int) async throws -> String {
        String(decoding: try await readRaw(count), as: UTF8.self)
    }

    /// Exactly `count` bytes, undecoded.
    ///
    /// A body is not text until something has read the charset it
    /// declares, and that header is inside the bytes. Decoding here
    /// would settle the question before it has been asked.
    private func readRaw(_ count: Int) async throws -> Data {
        while buffer.count < count {
            let chunk = try await receive()
            if chunk.isEmpty { throw Failure.closed }
            buffer.append(chunk)
        }
        let body = buffer.prefix(count)
        buffer.removeFirst(count)
        return Data(body)
    }

    /// One message, whole.
    ///
    /// `BODY.PEEK[]` rather than `BODY[]` for the same reason the
    /// header fetch peeks: opening a message is the reader's decision,
    /// and a client that marks mail read for having looked at it takes
    /// that decision away. Marking read is a separate, deliberate call.
    func fetchRaw(uid: UInt32) async throws -> Data {
        let t = nextTag()
        try await send("\(t) UID FETCH \(uid) (BODY.PEEK[])\r\n")
        var out = Data()
        while true {
            let line = try await readLine()
            if let done = IMAP.completion(of: line, tag: t) {
                switch done {
                case .ok: return out
                case let .no(d): throw Failure.server(d)
                case let .bad(d): throw Failure.server(d)
                }
            }
            guard let announced = IMAP.fetchLine(line), let count = announced.literalBytes
            else { continue }
            // The announced byte count, never a scan for a terminator:
            // a message contains every byte sequence a terminator could
            // be made of.
            out = try await readRaw(count)
        }
    }

    /// Mark a message read on the server, because somebody read it.
    func markSeen(uid: UInt32) async throws {
        try await store(uid: uid, op: "+FLAGS", flag: "\\Seen")
    }

    /// Mark it unread again.
    ///
    /// `-FLAGS`, not a `FLAGS` that names what should remain: the
    /// second replaces the whole set, so it would quietly clear
    /// `\Flagged`, `\Answered` and every keyword the person or another
    /// client had put there.
    func markUnseen(uid: UInt32) async throws {
        try await store(uid: uid, op: "-FLAGS", flag: "\\Seen")
    }

    private func store(uid: UInt32, op: String, flag: String) async throws {
        let t = nextTag()
        try await send("\(t) UID STORE \(uid) \(op) (\(flag))\r\n")
        try await awaitCompletion(t)
    }

    /// What the server said it can do.
    ///
    /// Read rather than assumed: the two commands below exist only on
    /// some servers, and asking for one that is not there is an error
    /// the person sees rather than a fallback they do not.
    func capabilities() async throws -> Set<String> {
        let (untagged, done) = try await command("CAPABILITY")
        if case let .no(d) = done { throw Failure.server(d) }
        if case let .bad(d) = done { throw Failure.server(d) }
        var out: Set<String> = []
        for line in untagged {
            if case let .capabilities(names) = line {
                for name in names { out.insert(name.uppercased()) }
            }
        }
        return out
    }

    /// Put a message in another folder.
    ///
    /// `MOVE` (RFC 6851) where the server has it, and the older
    /// three-step dance where it does not — and the difference matters
    /// more than it looks:
    ///
    /// **A bare `EXPUNGE` removes every message in the folder flagged
    /// `\Deleted`**, including ones somebody else's client flagged and
    /// has not expunged yet. `UID EXPUNGE` (RFC 4315) removes only the
    /// one named. Where neither `MOVE` nor `UIDPLUS` is offered, the
    /// message is flagged and **left** rather than expunged: it
    /// disappears from the list either way, and no other message is
    /// taken with it.
    func moveTo(uid: UInt32, folder: String, capabilities: Set<String>) async throws {
        if capabilities.contains("MOVE") {
            let t = nextTag()
            try await send("\(t) UID MOVE \(uid) \(IMAP.quoted(folder))\r\n")
            try await awaitCompletion(t)
            return
        }
        let copy = nextTag()
        try await send("\(copy) UID COPY \(uid) \(IMAP.quoted(folder))\r\n")
        try await awaitCompletion(copy)
        try await store(uid: uid, op: "+FLAGS", flag: "\\Deleted")
        if capabilities.contains("UIDPLUS") {
            let expunge = nextTag()
            try await send("\(expunge) UID EXPUNGE \(uid)\r\n")
            try await awaitCompletion(expunge)
        }
    }

    private func awaitCompletion(_ tag: String) async throws {
        while true {
            let line = try await readLine()
            if let done = IMAP.completion(of: line, tag: tag) {
                switch done {
                case .ok: return
                case let .no(d): throw Failure.server(d)
                case let .bad(d): throw Failure.server(d)
                }
            }
        }
    }

    // MARK: - the wire

    private func nextTag() -> String {
        tag += 1
        return "a\(tag)"
    }

    /// Send a command and read until its tagged reply.
    private func command(_ text: String) async throws -> ([IMAP.Untagged], IMAP.Completion) {
        let t = nextTag()
        try await send("\(t) \(text)\r\n")
        var untagged: [IMAP.Untagged] = []
        while true {
            let line = try await readLine()
            if let done = IMAP.completion(of: line, tag: t) {
                return (untagged, done)
            }
            if let u = IMAP.untagged(line) {
                untagged.append(u)
            }
        }
    }

    private func send(_ text: String) async throws {
        do {
            try await transport.send(Data(text.utf8))
        } catch let e as TransportFailure {
            throw failure(e)
        }
    }

    /// One CRLF-terminated line.
    ///
    /// Buffered, because a read returns whatever arrived rather than
    /// one line: two replies can land in a single packet, and half a
    /// reply can land in two.
    private func readLine() async throws -> String {
        while true {
            if let r = buffer.firstRange(of: Data("\r\n".utf8)) {
                let line = String(decoding: buffer[..<r.lowerBound], as: UTF8.self)
                buffer.removeSubrange(..<r.upperBound)
                return line
            }
            let chunk = try await receive()
            if chunk.isEmpty { throw Failure.closed }
            buffer.append(chunk)
        }
    }

    private func receive() async throws -> Data {
        do {
            return try await transport.receive()
        } catch let e as TransportFailure {
            throw failure(e)
        }
    }
}
